From face05999065c995c660b822e23b32ee723637fd Mon Sep 17 00:00:00 2001
From: Vignesh Sridharan <vignsrid@qti.qualcomm.com>
Date: Tue, 21 Jul 2026 12:56:03 +0530
Subject: [PATCH] agent: Populate SMD MAC in Agent AP MLD TLV for TD response

Store the SMD MAC received in M2 into AgentDB::sAPMLDConfiguration
and include it in the affiliated_ap_smd_mac_addr field when building
the Agent AP MLD Configuration TLV in the Topology Discovery response.
---
 agent/src/beerocks/slave/agent_db.h                       | 1 +
 agent/src/beerocks/slave/son_slave_thread.cpp             | 8 ++++++++
 .../beerocks/slave/tasks/ap_autoconfiguration_task.cpp    | 3 +++
 3 files changed, 12 insertions(+)

diff --git a/agent/src/beerocks/slave/agent_db.h b/agent/src/beerocks/slave/agent_db.h
index fc66e608a..f413c6908 100644
--- a/agent/src/beerocks/slave/agent_db.h
+++ b/agent/src/beerocks/slave/agent_db.h
@@ -930,6 +930,7 @@ public:
 
         sMLDConfiguration mld_config;
         std::vector<sAffiliatedAP> affiliated_aps;
+        sMacAddr smd_mac = net::network_utils::ZERO_MAC; ///< SMD MAC received from controller in M2
     } sAPMLDConfiguration;
 
     typedef struct {
diff --git a/agent/src/beerocks/slave/son_slave_thread.cpp b/agent/src/beerocks/slave/son_slave_thread.cpp
index 1df2e01f6..6cd56ce4f 100644
--- a/agent/src/beerocks/slave/son_slave_thread.cpp
+++ b/agent/src/beerocks/slave/son_slave_thread.cpp
@@ -7821,6 +7821,14 @@ bool slave_thread::add_agent_ap_mld_configuration_tlv(ieee1905_1::CmduMessageTx
                 ap_mld->modes().emlmr = 1;
             }
 
+            // Populate SMD MAC address if stored during cloning (M2 parsing)
+            if (ap_mld_conf.smd_mac != net::network_utils::ZERO_MAC) {
+                ap_mld->affiliated_ap_smd_mac_addr() = ap_mld_conf.smd_mac;
+                ap_mld->modes().affiliated_ap_smd_mac_addr_valid = 1;
+                LOG(DEBUG) << "TD response: set SMD MAC " << ap_mld_conf.smd_mac
+                           << " for SSID " << ap_mld_conf.mld_config.mld_ssid;
+            }
+
             LOG(DEBUG) << "Sending AP MLD configuration for " << ap_mld_conf.mld_config.mld_ssid
                        << " [mac=" << ap_mld_conf.mld_config.mld_mac << ", mode=" << std::hex
                        << ap_mld_conf.mld_config.mld_mode << "]";
diff --git a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
index 24ede6033..4d4585a38 100644
--- a/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
+++ b/agent/src/beerocks/slave/tasks/ap_autoconfiguration_task.cpp
@@ -3378,6 +3378,9 @@ bool ApAutoConfigurationTask::handle_agent_ap_mld_configuration_tlv(
                 LOG(DEBUG) << "SMD MAC for SSID " << ssid << ": " << smd_mac;
             }
 
+            // Persist SMD MAC into AgentDB so it is available for the TD response
+            current_ap_mld_conf.smd_mac = smd_mac;
+
             m_ap_mld_requests_infos[rad_iface][current_ap_mld_conf.mld_config.mld_ssid] = {
                 current_ap_mld_conf.mld_config.mld_unit, current_ap_mld_conf.mld_config.mld_mode,
                 smd_mac};
-- 
2.34.1

